packages/plugins/response-cache/src/plugin.ts

```diff
@@ -195,41 +181,106 @@ export function useResponseCache({
-      if (schema.extensions?.[appliedTransform] === true) {
-        return;
+    async onExecute(onExecuteParams) {
+      let documentChanged = false;
+      const newDocument = visit(onExecuteParams.args.document, {
+        SelectionSet(node): SelectionSetNode {
+          if (!node.selections.some(selection => selection.kind === Kind.FIELD && selection.name.value === '__typename')) {
+            documentChanged = true;
+            return {
+              ...node,
+              selections: [
+                {
+                  kind: Kind.FIELD,
+                  name: {
+                    kind: Kind.NAME,
+                    value: '__typename',
+                  },
+                },
+                ...node.selections,
+              ],
+            };
+          }
+          return node;
+        },
+      });
+      if (documentChanged) {
+        onExecuteParams.setExecuteFn(function typeNameAddedExecute() {
+          return onExecuteParams.executeFn({
+            ...onExecuteParams.args,
+            document: newDocument,
+          });
+        });
       }
-
-      const patchedSchema = applyResponseCacheLogic(schema, idFields);
-      patchedSchema.extensions = {
-        ...patchedSchema.extensions,
-        [appliedTransform]: true,
-      };
-      replaceSchema(patchedSchema);
-    },
-    async onExecute(ctx) {
       const identifier = new Map<string, CacheEntityRecord>();
       const types = new Set<string>();
 
-      const context: Context = {
-        identifier,
-        types,
-        currentTtl: undefined,
-        ttlPerType,
-        ignoredTypesMap,
-        skip: false,
-      };
-
-      ctx.extendContext({
-        [contextSymbol]: context,
-      });
+      let currentTtl: number | undefined;
+      let skip = false;
+
+      const processResult = (result: ExecutionResult): ExecutionResult =>
+        visitResult(
+          result,
+          {
+            document: onExecuteParams.args.document,
+            variables: onExecuteParams.args.variableValues as any,
+            operationName: onExecuteParams.args.operationName ?? undefined,
+            rootValue: onExecuteParams.args.rootValue,
+            context: onExecuteParams.args.contextValue,
+          },
+          onExecuteParams.args.schema,
+          new Proxy(
+            {},
+            {
+              get(_, typename: string) {
+                let typenameCalled = 0;
+                return new Proxy((val: any) => val, {
+                  // Needed for leaf values
+                  apply(_, __, [val]) {
+                    return val;
+                  },
+                  get(_, fieldName: string) {
+                    if (documentChanged) {
+                      if (fieldName === '__typename') {
+                        typenameCalled++;
+                      }
+                      if (fieldName === '__leave') {
+                        // When typename is defined in the selection set, it calls it twice.
```

Comment:
I think we need some inline documentation on why this is the case, I am very confused.

Reply:
I'll add more details with references from graphql-tools source.